pubs/bibstruct.py

```diff
@@ -56,26 +65,51 @@ def valid_citekey(citekey):
     # FIXME: a bit crude, but efficient for now (and allows unicode citekeys)
     return not '/' in citekey
 
-
-def generate_citekey(bibdata):
+class CitekeyFormatter(Formatter):
+    def __init__(self):
+        super(CitekeyFormatter, self).__init__()
+
+    def format_field(self, val, fmt):
+        if len(fmt) > 0 and fmt[0] == 'u':
+            s = str(val).upper()
+            fmt = fmt[1:]
+        elif len(fmt) > 0 and fmt[0] == 'l':
+            s = str(val).lower()
+            fmt = fmt[1:]
+        else:
+            s = val
+        return str2citekey(s.__format__(fmt))
+
+
+    def get_value(self, key, args, entry):
```

Comment:
One likely problem: content does not get sanitized. So names with latex in them may not produce legal citekeys (i.e. "Le~Clerc de La~Th{\'e}venaz"). Sanitizing here might not be the best way though, as it might be hard. Better would be to check and gracefully fail, and inform the user that she has to edit the citekey herself. In any case, usecases are the way to go to test robustness against those.